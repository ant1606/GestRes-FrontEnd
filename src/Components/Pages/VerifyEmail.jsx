import React, {useEffect} from 'react';
import {useNavigate, useParams} from "react-router-dom";
import useSecurity from "../../Context/SecurityContext.jsx";

const VerifyEmail = () => {
    let {id, hash} = useParams();
    const {verifyUserEmail} = useSecurity();
    const navigate = useNavigate();

    useEffect( ()=>{
        async function verifiedUserEmail(id, hash) {
            let res = await verifyUserEmail(id, hash);
            if(res){
                navigate("/dashboard");
            }
        }

        verifiedUserEmail(id, hash);
    },[]);



    return (
        <>
        <div>Verificando email</div>
        </>
    );
}
export default VerifyEmail;