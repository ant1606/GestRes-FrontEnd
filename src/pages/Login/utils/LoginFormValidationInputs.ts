import { type LoginFormData, type ValidationInputMessage } from '../index.types';

export const validateUserEmail = (values: LoginFormData): ValidationInputMessage => {
  const emailToValidate = values.email.trim();
  const regExFormatEmail = /^[\w-\\.]+@([\w-]+\.)+[\w-]{2,4}$/;
  const emailValidFormat = regExFormatEmail.test(emailToValidate);
  if (emailToValidate.length === 0) {
    return 'Debe ingresar el email del usuario';
  }
  if (!emailValidFormat) {
    return 'Formato incorrecto ingresado del email';
  }
  return null;
};

export const validateUserPassword = (values: LoginFormData): ValidationInputMessage => {
  const passwordToValidate = values.password.trim();
  if (passwordToValidate.length === 0) {
    return 'Debe ingresar el password del usuario';
  }
  return null;
};
