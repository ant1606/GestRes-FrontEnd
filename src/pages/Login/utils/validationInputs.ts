import { type LoginFormData } from '../index.types';

export const validateUserEmail = (values: LoginFormData): ValidationInputResult => {
  const emailToValidate = values.email.trim();
  const regExFormatEmail = /^[\w-\\.]+@([\w-]+\.)+[\w-]{2,4}$/;
  const emailValidFormat = regExFormatEmail.test(emailToValidate);
  if (emailToValidate.length === 0) {
    return 'Debe ingresar el email del usuario';
  }
  if (!emailValidFormat) {
    return 'Formato incorrecto ingresado del email';
  }
  return null;
};

export const validateUserPassword = (values: LoginFormData): ValidationInputResult => {
  const passwordToValidate = values.password.trim();
  if (passwordToValidate.length === 0) {
    return 'Debe ingresar el password del usuario';
  }
  return null;
};
