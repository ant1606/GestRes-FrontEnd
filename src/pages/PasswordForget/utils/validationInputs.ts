import { type PasswordForgetFormData } from '../index.types';

export const validateUserEmail = (values: PasswordForgetFormData): ValidationInputResult => {
  const emailToValidate = values.email.trim();
  const regExFormatEmail = /^[\w-\\.]+@([\w-]+\.)+[\w-]{2,4}$/;
  const emailValidFormat = regExFormatEmail.test(emailToValidate);
  if (emailToValidate.trim().length === 0) {
    return 'Debe ingresar el email del usuario';
  }
  if (!emailValidFormat) {
    return 'Formato incorrecto ingresado del email';
  }
  return null;
};
