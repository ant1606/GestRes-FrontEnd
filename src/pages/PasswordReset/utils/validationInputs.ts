import { type ResetPasswordFormData } from '../index.types';

export const validateUserEmail = (values: ResetPasswordFormData): ValidationInputResult => {
  const emailToValidate = values.email.trim();
  const regExFormatEmail = /^[\w-\\.]+@([\w-]+\.)+[\w-]{2,4}$/;
  const emailValidFormat = regExFormatEmail.test(emailToValidate);
  if (emailToValidate.trim().length === 0) {
    return 'Debe ingresar el email del usuario';
  }
  if (!emailValidFormat) {
    return 'Formato incorrecto ingresado del email';
  }
  return null;
};

export const validateUserPassword = (values: ResetPasswordFormData): ValidationInputResult => {
  const passwordToValidate = values.password?.trim();
  if (passwordToValidate.trim().length === 0) {
    return 'Debe ingresar el password del usuario';
  }
  return null;
};

export const validateUserPasswordConfirmation = (
  values: ResetPasswordFormData
): ValidationInputResult => {
  const passwordConfirmationToValidate = values.passwordConfirmation?.trim();
  const passwordToValidate = values.password?.trim();
  if (passwordConfirmationToValidate.trim().length === 0) {
    return 'Debe ingresar la confirmación de password del usuario';
  }
  if (passwordConfirmationToValidate !== passwordToValidate) {
    return 'Las contraseñas ingresadas no son iguales';
  }
  return null;
};
