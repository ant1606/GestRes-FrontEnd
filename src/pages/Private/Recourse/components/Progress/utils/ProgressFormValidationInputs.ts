import moment from 'moment';
import { type ProgressFormData } from '../indext.types';

export const validateAdvancedAmount = (values: ProgressFormData): ValidationInputResult => {
  if (values.advanced < 1) {
    return 'El Avance  realizado debe ser mayor a 0';
  }
  if (values.lastProgress !== undefined && values.advanced <= values.lastProgress.advanced) {
    return `El Avance  realizado debe ser mayor al anterior avance realizado que fue  ${values.lastProgress.advanced}`;
  }
  if (values.lastProgress !== undefined && values.advanced > values.lastProgress.total) {
    return `Avance debe ser menor o igual a la cantidad Total del Recurso que es ${values.lastProgress.total}`;
  }
  return null;
};

export const validateDoneAmount = (values: ProgressFormData): ValidationInputResult => {
  // if (values.done < 1) {
  //   return 'Avance debe ser mayor a 0';
  // }
  // if (values.lastProgress !== undefined && values.done > values.lastProgress.pending) {
  //   return 'Avance debe ser menor a Pendiente';
  // }
  return null;
};

export const validateDate = (values: ProgressFormData): ValidationInputResult => {
  const myDate = moment(values.date);
  const dateOfLastStatus = moment(values.lastProgress?.date);
  const differenceBetweenDays = myDate.diff(dateOfLastStatus, 'days');
  if (differenceBetweenDays < 0)
    return 'No puede registrar una fecha anterior a la Ãºltima fecha del registro de estado del recurso';
  return null;
};

export const validateComment = (): ValidationInputResult => {
  return null;
};
