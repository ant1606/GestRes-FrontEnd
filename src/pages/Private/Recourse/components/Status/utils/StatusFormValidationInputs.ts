import moment from 'moment';
import { type StatusFormData } from '../index.types';
import { GLOBAL_STATUS_RECOURSE } from '@/config/globalConstantes';

export const validateFecha = (values: StatusFormData): ValidationInputResult => {
  console.log(values);
  const myDate = moment(values.date);
  const dateOfLastStatus = moment(values.lastStatusOfRecourse?.date);
  const differenceBetweenDays = myDate.diff(dateOfLastStatus, 'days');
  if (differenceBetweenDays < 0)
    return 'No puede registrar una fecha anterior a la Ãºltima fecha del registro de estado del recurso';
  return null;
};

export const validateEstadoId = (values: StatusFormData): ValidationInputResult => {
  return null;
};

export const validateComentario = (values: StatusFormData): ValidationInputResult => {
  const stateName = values.recourseStatus?.filter((status) => status.id === values.statusId)[0]
    .value;
  if (
    stateName === GLOBAL_STATUS_RECOURSE.DESCARTADO ||
    stateName === GLOBAL_STATUS_RECOURSE.DESFASADO
  ) {
    if (values.comment.trim() === '') {
      return `Debe ingresar un comentario al cambiar a estado ${stateName}`;
    }
  }
  return null;
};
