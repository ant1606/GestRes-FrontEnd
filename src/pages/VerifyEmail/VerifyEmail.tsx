import React, { useEffect } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import AuthenticationTemplate from '../../components/AuthenticationTemplate.js';
import { useAppDispatch } from '#/hooks/redux/index.js';
import { isLoading } from '#/redux/slice/uiSlice.js';
import { userIsLoggin } from '#/redux/slice/authenticationSlice.js';
import { verifyUserEmail } from '#/services/verifyEmail.services.js';
import { toastNotifications } from '#/utilities/notificationsSwal.js';
import { savePersistenDataUser } from '#/utilities/authenticationManagement.js';
import { type VerifyEmailErrorResponse, type VerifyEmailSuccessResponse } from './index.types.js';

export const VerifyEmail: React.FC = () => {
  const { id, hash } = useParams();
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  useEffect(() => {
    verifyEmail();
  });

  const userIsLoggedInPromise = async (params: User): Promise<void> => {
    await new Promise<void>((resolve) => {
      dispatch(
        userIsLoggin({
          email: params.email,
          id: parseInt(params.id),
          isVerified: params.isVerified,
          name: params.name
        })
      );
      resolve();
    });
  };

  const verifyEmail = async (): Promise<void> => {
    try {
      dispatch(isLoading(true));
      if (id?.trim() === '' || id === '0' || id === undefined)
        throw new Error('No se identifico al usuario');
      if (hash?.trim() === '' || hash === '' || hash === undefined)
        throw new Error('No se identifico al usuario');

      const response = await verifyUserEmail(id, hash);

      if (response.status === 'error') {
        const responseError = response as VerifyEmailErrorResponse;

        // Mensaje de error general por parte del backend
        if (responseError.message !== '') {
          throw new Error(responseError.message);
        }
      } else {
        const responseError = response as VerifyEmailSuccessResponse;
        savePersistenDataUser(responseError.data);
        await userIsLoggedInPromise(responseError.data.user);
        navigate('/app/dashboard', { replace: true });
      }
    } catch (error: any) {
      toastNotifications().notificationError(error.message);
    } finally {
      dispatch(isLoading(false));
    }
  };

  return (
    <>
      <AuthenticationTemplate>
        <p className="text-5xl leading-10 font-bold text-center">
          Estamos verificando tu correo electr√≥nico ...
        </p>
      </AuthenticationTemplate>
    </>
  );
};
