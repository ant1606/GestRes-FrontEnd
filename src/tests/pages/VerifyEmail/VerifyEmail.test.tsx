import { VerifyEmail } from '@/pages/VerifyEmail/VerifyEmail';
import { setServiceVerifyEmailResponseSuccess } from '@/tests/mocks/verifyEmail.handlers';
import { renderWithProviders } from '@/tests/utils/renderWithProvider';
import { cleanup, waitFor } from '@testing-library/react';
import Cookies from 'js-cookie';
import { Route, Routes } from 'react-router-dom';

describe('Test en VerifyEmail', () => {
  beforeEach(() => {
    cleanup();
    setServiceVerifyEmailResponseSuccess(true);
  });

  test('VerifyEmail debe renderizarse', () => {
    const wrapper = renderWithProviders(
      <Routes>
        <Route path="/verifyEmail/:id/:hash" element={<VerifyEmail />} />
      </Routes>,
      undefined,
      ['/verifyEmail/1/miHash']
    );
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  test('Debe lanzar error si el id en la ruta es 0, vacio no es valido', async () => {
    const id = ' ';
    const hash = 'miHash';
    const wrapper = renderWithProviders(
      <Routes>
        <Route path="/verifyEmail/:id/:hash" element={<VerifyEmail />} />
        {/* <Route path="*" element={<>Pagina no existe</>} /> */}
      </Routes>,
      undefined,
      [`/verifyEmail/${id}/${hash}`]
    );
    await waitFor(() => {
      expect(wrapper.getByText('No se identifico al usuario')).toBeInTheDocument();
    });
    wrapper.unmount();
  });

  test('Debe lanzar error si el hash en la ruta es vacio o no valido', async () => {
    const id = '123';
    const hash = ' ';
    const wrapper = renderWithProviders(
      <Routes>
        <Route path="/verifyEmail/:id/:hash" element={<VerifyEmail />} />
        {/* <Route path="*" element={<>Pagina no existe</>} /> */}
      </Routes>,
      undefined,
      [`/verifyEmail/${id}/${hash}`]
    );
    await waitFor(() => {
      expect(wrapper.getByText('No se identifico al usuario')).toBeInTheDocument();
    });
    wrapper.unmount();
  });

  test('Debe mostrar una notificación si se obtiene un error por parte de la API ', async () => {
    setServiceVerifyEmailResponseSuccess(false);
    const id = '123';
    const hash = 'miHash';
    const wrapper = renderWithProviders(
      <Routes>
        <Route path="/verifyEmail/:id/:hash" element={<VerifyEmail />} />
        {/* <Route path="*" element={<>Pagina no existe</>} /> */}
      </Routes>,
      undefined,
      [`/verifyEmail/${id}/${hash}`]
    );
    await waitFor(() => {
      expect(
        wrapper.getByText('Hubo problemas en la comunicación con el servidor')
      ).toBeInTheDocument();
    });
    wrapper.unmount();
  });

  test('Debe mostrar app/dashboard si se verifica satisfactoriamente el email del usuario', async () => {
    const id = '123';
    const hash = 'miHash';
    setServiceVerifyEmailResponseSuccess(true);
    const wrapper = renderWithProviders(
      <Routes>
        <Route path="/verifyEmail/:id/:hash" element={<VerifyEmail />} />
        <Route path="/app/dashboard" element={<>Dashboard</>} />
      </Routes>,
      undefined,
      [`/verifyEmail/${id}/${hash}`]
    );
    await waitFor(() => {
      expect(wrapper.getByText('Dashboard')).toBeInTheDocument();
    });
    wrapper.unmount();
  });

  test('Debe almacenar en cookie y en el store la información del usuario si se verifica satisfactoriamente el email del usuario', async () => {
    const id = '123';
    const hash = 'miHash';
    setServiceVerifyEmailResponseSuccess(true);
    const wrapper = renderWithProviders(
      <Routes>
        <Route path="/verifyEmail/:id/:hash" element={<VerifyEmail />} />
        <Route path="/app/dashboard" element={<>Dashboard</>} />
      </Routes>,
      undefined,
      [`/verifyEmail/${id}/${hash}`]
    );
    await waitFor(() => {
      expect(Cookies.get('bearerToken')).toBe('miToken');
      expect(localStorage.getItem('user')).not.toBe(undefined);
    });
    wrapper.unmount();
  });
});
