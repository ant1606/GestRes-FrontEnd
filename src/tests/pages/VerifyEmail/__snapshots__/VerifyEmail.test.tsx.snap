// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test en VerifyEmail > VerifyEmail debe renderizarse 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col h-screen"
      >
        <div
          class="bg-gray-800 min-h-[5rem]"
        />
        <div
          class="bg-gray-200 h-full flex justify-center align-center"
        >
          <div
            class="flex flex-col justify-center align-center gap-6 min-w-[30rem]"
          >
            <p
              class="text-5xl leading-10 font-bold text-center"
            >
              Estamos verificando tu correo electrónico ...
            </p>
          </div>
        </div>
        <div
          class="bg-gray-900 min-h-[5rem]"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col h-screen"
    >
      <div
        class="bg-gray-800 min-h-[5rem]"
      />
      <div
        class="bg-gray-200 h-full flex justify-center align-center"
      >
        <div
          class="flex flex-col justify-center align-center gap-6 min-w-[30rem]"
        >
          <p
            class="text-5xl leading-10 font-bold text-center"
          >
            Estamos verificando tu correo electrónico ...
          </p>
        </div>
      </div>
      <div
        class="bg-gray-900 min-h-[5rem]"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
